### PR TITLE
fix: authz-keycloak plugin should not be marked as auth-type and should run in access phase, basic-auth plugin should run in rewrite phase.

### PR DESCRIPTION
**Pull request by jenskeiner**
### What this PR does / why we need it:
See #2857.

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [ ] Have you added corresponding test cases?
* [x] Have you modified the corresponding document?
* [x] Is this PR backward compatible? **If it is not backward compatible, please discuss on the [mailing list](https://github.com/apache/apisix/tree/master#community) first**

### CODE REVIEW

**Comment by membphis** [comment]
Authentication type plug-ins should run in the rewrite phase

**Comment by jenskeiner** [comment]
> Authentication type plug-ins should run in the rewrite phase

How do you suggest this get's fixed then when `openid-connect` and other authentication plugins currently run in the access phase?

The plugin development guidelines are currently not in line with the actual code base.

Also, what's the rationale for running authentication/authorization in the *rewrite* phase when the *access* phase is the one that specifically caters for that purpose? It's not obvious why you would run in a different phase.

**Comment by membphis** [comment]
reply you later

need more time to read your issue and pr first

**Comment by spacewander** [comment]
@membphis @jenskeiner 
`authz-keycloak` is for authorization but not authentication. It doesn't interact with consumers, so I think this change is fine.

File: doc/plugin-develop.md [hunk 1]
```diff
@@ -168,8 +168,8 @@ local _M = {
 Determine which phase to run, generally access or rewrite. If you don't know the [Openresty life cycle](https://openresty-reference.readthedocs.io/en/latest/Directives/), it's
 recommended to know it in advance. For example key-auth is an authentication plugin, thus the authentication should be completed
 before forwarding the request to any upstream service. Therefore, the plugin can be executed in the rewrite and access phases.
-In APISIX, the authentication logic is implemented in the rewrite phase. Generally, IP access and interface
-permission are completed in the access phase.
+In APISIX, the authentication logic is implemented in the access phase. Generally, IP access and interface
```
Comment (spacewander) [inline doc/plugin-develop.md:171]: We don't need to change it as authz-keycloak is for authorization.

**Comment by membphis** [comment]
> @membphis @jenskeiner
> `authz-keycloak` is for authorization but not authentication. It doesn't interact with consumers, so I think this change is fine.

many thx for your explain. @spacewander 
we can continue this PR

File: doc/plugin-develop.md (hunk 1 above)
Comment (jenskeiner) [reply doc/plugin-develop.md:171]: Ok, I can revert this one. However, note that some authentication plugins currently run in the access phase, not the rewrite one. So the document doesn’t reflect the current state correctly, I believe.

Comment (spacewander) [reply doc/plugin-develop.md:171]: @jenskeiner
We can fix it later, but in those plugins, not the documentation.

Comment (spacewander) [reply doc/plugin-develop.md:171]: @jenskeiner
Thank you for the problem you point out. I just search for all plugins which has `type = 'auth'` (excluding the authz-keycloak) and find out `basic-auth` doesn't run the phase correctly.

Comment (spacewander) [reply doc/plugin-develop.md:171]: The reason why we need to run authentication plugins in the rewrite phases:
https://github.com/apache/apisix/blob/f4161d39a495879edffbc75e2099f6e487bae81d/apisix/init.lua#L508-L520
And:
https://github.com/apache/apisix/blob/f4161d39a495879edffbc75e2099f6e487bae81d/apisix/admin/consumers.lua#L50

Plugins has `type = 'auth'` should run in rewrite phase and set up the consumer for current api_ctx.

Comment (jenskeiner) [reply doc/plugin-develop.md:171]: Ok, understood why you need to run in the rewrite phase. Then I would propose to change the phase for all authentication-type plugins to *rewrite*, instead of adjusting the phase for `authz-keycloak` from *rewrite* to *access*. Also, I think it would make the plugin development guidelines clearer if the reason they need to run in the rewrite phase was pointed out more clearly.

If you're fine with this, I can update this PR accordingly. Just let me know what you think.

File: apisix/plugins/basic-auth.lua [hunk 2]
```diff
@@ -125,7 +125,7 @@ do
     end
 end
 
-function _M.access(conf, ctx)
+function _M.rewrite(conf, ctx)
```
Comment (spacewander) [inline apisix/plugins/basic-auth.lua:128]: Need a test case to check if the `ctx.consumer` set by this plugin can be used by other plugin runs in access phase.

File: doc/plugin-develop.md (hunk 1 above)
Comment (spacewander) [inline doc/plugin-develop.md:171]: Remember to revert this change.

Comment (jenskeiner) [reply doc/plugin-develop.md]: > We don't need to change it as authz-keycloak is for authorization.

Reverted.

File: apisix/plugins/basic-auth.lua (hunk 2 above)
Comment (jenskeiner) [reply apisix/plugins/basic-auth.lua:128]: Happy to add a test case. I don't have much experience writing them yet, so if you could point to a similar example, that would be helpful.

File: doc/plugin-develop.md (hunk 1 above)
Comment (jenskeiner) [reply doc/plugin-develop.md]: Done.

File: apisix/plugins/basic-auth.lua (hunk 2 above)
Comment (spacewander) [reply apisix/plugins/basic-auth.lua:128]: You can add a stub:
```diff
diff --git apisix/init.lua apisix/init.lua
index 7333dd1..411cdd4 100644
--- apisix/init.lua
+++ apisix/init.lua
@@ -514,6 +514,10 @@ function _M.http_access_phase()
                 api_ctx.consumer,
                 api_ctx
             )
+
+            core.log.info("find consumer ", api_ctx.consumer.username,
+                          ", config changed: ", changed)
+
             if changed then
                 core.table.clear(api_ctx.plugins)
                 api_ctx.plugins = plugin.filter(route, api_ctx.plugins)
```

And then check it in:
https://github.com/apache/apisix/blob/e66dd9a314e0b545e412f806b8b8d98f3b1a8f2d/t/plugin/basic-auth.t#L192

Add something like:
```
--- error_log
find consumer foo
```

Comment (jenskeiner) [reply apisix/plugins/basic-auth.lua:128]: Ok, I think I've seen an example in the test cases for `key-auth`. I'll try to add test cases acceding for `basic-auth`.

Comment (membphis) [inline apisix/plugins/basic-auth.lua:128]: why do we need to change another plugin?

one PR for one thing, we should only update something about plugin `authz-keycloak`

please revert this code

File: apisix/plugins/openid-connect.lua [hunk 3]
```diff
@@ -49,6 +49,7 @@ local schema = {
 local _M = {
     version = 0.1,
     priority = 2599,
+    type = 'auth',
```
Comment (membphis) [inline apisix/plugins/openid-connect.lua:52]: ditto

you can create a new PR to fix this

**Review by membphis — CHANGES_REQUESTED** [review]

File: apisix/plugins/openid-connect.lua (hunk 3 above)
Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:52]: I'm sorry but I currently can't afford the time to split this into separate PRs. I think the change set is reasonably small to be considered in one PR. If you prefer more fine-grained PRs, then please consider creating these yourself, e.g. based on the changes in this branch.

Comment (membphis) [reply apisix/plugins/openid-connect.lua:52]: I found you have submitted new PR: https://github.com/apache/apisix/pull/2905/files

Many thanks for your nice job.

Small PR is important, it easy to find problems and we can merge it asap when it is small.

By the way, it is a clearer way to show what we are doing.

Many thx for nice contribution again @jenskeiner

Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:52]: Sure, you're very welcome. Thanks for the info. I'll try to put the remaining outstanding changes from this one into a separate new PR and then we take it from there.

Comment (jenskeiner) [reply apisix/plugins/openid-connect.lua:52]: Have created #2910 which should contain the outstanding changes from this PR.

**Comment by jenskeiner** [comment]
Closing this one as it has since been superseded by other PRs addressing the same issues.
